cloudnet-driver/src/main/java/de/dytanic/cloudnet/driver/module/IModuleProvider.java

```diff
@@ -20,53 +20,229 @@
     this.setModuleDirectoryPath(moduleDirectory.toPath());
   }
 
-  Path getModuleDirectoryPath();
-
-  void setModuleDirectoryPath(Path moduleDirectory);
-
-  IModuleProviderHandler getModuleProviderHandler();
-
-  void setModuleProviderHandler(IModuleProviderHandler moduleProviderHandler);
-
-  IModuleDependencyLoader getModuleDependencyLoader();
-
-  void setModuleDependencyLoader(IModuleDependencyLoader moduleDependencyLoader);
-
-  Collection<IModuleWrapper> getModules();
-
-  Collection<IModuleWrapper> getModules(String group);
-
-  IModuleWrapper getModule(String name);
-
-  IModuleWrapper loadModule(URL url);
-
-  IModuleWrapper loadModule(File file);
-
-  IModuleWrapper loadModule(Path path);
-
-  IModuleProvider loadModule(URL... urls);
-
-  IModuleProvider loadModule(File... files);
-
-  IModuleProvider loadModule(Path... paths);
+  /**
+   * Get the module provider handler of this provider or {@code null} when no handler is specified.
+   *
+   * @return the module provider handler of this provider or {@code null}.
+   * @see #setModuleProviderHandler(IModuleProviderHandler)
+   */
+  @Nullable IModuleProviderHandler getModuleProviderHandler();
+
+  /**
+   * Sets the module provider handler of this provider.
+   *
+   * @param moduleProviderHandler the new module provider to use or {@code null} when no handler should be used.
+   * @see #getModuleProviderHandler()
+   */
+  void setModuleProviderHandler(@Nullable IModuleProviderHandler moduleProviderHandler);
+
+  /**
+   * Get the module dependency loader. It's used to load all dependencies of all modules.
+   * <p>This handler is by default {@link DefaultMemoryModuleDependencyLoader}.</p>
+   *
+   * @return the module provider handler used for this provider.
+   * @see #setModuleDependencyLoader(IModuleDependencyLoader)
+   * @see DefaultMemoryModuleDependencyLoader
+   * @see DefaultPersistableModuleDependencyLoader
+   */
+  @NotNull IModuleDependencyLoader getModuleDependencyLoader();
+
+  /**
+   * Sets the module dependency loader which should be used by this provider.
+   *
+   * @param moduleDependencyLoader the module dependency loader to use.
+   * @see #getModuleDependencyLoader()
+   * @see DefaultMemoryModuleDependencyLoader
+   * @see DefaultPersistableModuleDependencyLoader
+   */
+  void setModuleDependencyLoader(@NotNull IModuleDependencyLoader moduleDependencyLoader);
+
+  /**
+   * Get all loaded, started, stopped modules provided by this provider.
+   *
+   * @return all loaded, started, stopped modules provided by this provider.
+   * @see ModuleLifeCycle
+   */
+  @NotNull
+  @Unmodifiable Collection<IModuleWrapper> getModules();
+
+  /**
+   * Get all loaded, started, stopped modules provided by this provider which have the specific given group.
+   *
+   * @param group the group id of the modules to get.
+   * @return all loaded, started, stopped modules provided by this provider which have the specific given group.
```

Comment:
Should we perhaps inform that this collection is not modifiable, apart from the annotation?